### PR TITLE
[BROKEN] feat(elbv2): Supports new types of listener rule conditions

### LINKED ISSUE #3888: ALB: New types of listener rule conditions
## :rocket: Feature Request

Add listener rules to ALB

### General Information

* [x] :wave: I may be able to implement this feature request

* [ ] :warning: This feature might incur a breaking change

### Description

This should contain the posibilities to add rules based on different conditions.
Right now there is support for path and header and at least one is required.
It should be possible to allow on query, source-ip, request method as well

(https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-listenerrule-conditions.html)

### Proposed Solution

Not sure if this requires a breaking change or if it could be added with just new API.

### Environment

  - **CDK CLI Version:** 
  - **Module Version:** 
  - **OS:** 
  - **Language:**

### PR DESCRIPTION
**Pull request by karupanerura**
### Commit Message
feat(elbv2): Supports new types of listener rule conditions

Fixes #3888
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.expected.json [hunk 1]
```diff
@@ -943,9 +943,11 @@
         "Conditions": [
           {
             "Field": "path-pattern",
-            "Values": [
-              "a/b/c"
-            ]
+            "PathPatternConfig": {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.expected.json:946]: In order to not cause unnecessary changes in all our user's templates, can we make sure that if the configuration can be expressed in the "legacy" format, that we do so?

Otherwise everyone around the world will freak out if all their templates change after a CDK update.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts [hunk 2]
```diff
@@ -160,6 +160,227 @@ export interface RedirectResponse {
...
+ * Properties for a raw Cfn listener rule condition
+ */
+export interface RawListenerRuleConditionProps {
+  readonly [key: string]: any;
+}
+
+/**
+ * Interface for a listener rule condition
+ */
+export interface IListenerRuleCondition {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:173]: What you're doing here is good, but what we generally do is:

```ts
export [abstract] class ListenerRuleCondition {
  public static queryString(key, value): ListenerRuleCondition {
    return new ...;
  }

  public static hostHeader(values: string[]): ListenerRuleCondition {
    return new ...;
  }

  
  protected constructor(...) { ... }
}

// NOTE: not "export"ed.
class IfYouNeedThem extends ListenerRuleCondition {
}
```

So:

* (abstract) base class instead of interface. Whether it's abstract or not depends on how you want to model it. It's perfectly fine to make the class itself just a holder for a blob of JSON that it gets in the constructor (which is protected) and spits back out in the `render()` method for like 90% of the cases. If it turns out you need a specific subclass for some other case, you still can do that. Also still leaves the ability for customers to define their own subclasses.
* Static factory methods to create instances.
* (Generally) hide the specific implementation classes (if you have any).

This has the advantage that it's more discoverable. The declared type `conditions?: ListenerRuleCondition[]` will also be the type that you can just type into your IDE, hit `.`, and then see an autocompleted list of options. Otherwise you'll have to go digging around for all the classes you can `new` up, which is more work.

It's also fine to have two or more factory functions for the same type of condition, if their signatures are slightly different. Then you can make "simple" and convenient ones, and more complex ones, and don't have to expose the full complexity of the API to people who just need a simple condition.

Also wondering if we can just call this class `Condition` instead of `ListenerRuleCondition`, since people are going to have to type it a lot :/

Comment (rix0rrr) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:173]: Maybe also stick all the conditions in their own file.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts [hunk 3]
```diff
@@ -160,6 +160,227 @@ export interface RedirectResponse {
   readonly statusCode: string;
 }
 
+/**
+ * Properties for a raw Cfn listener rule condition
+ */
+export interface RawListenerRuleConditionProps {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:166]: Might as well just replace this with `any`, no?

File: packages/@aws-cdk/aws-elasticloadbalancingv2/README.md [hunk 4]
```diff
@@ -95,6 +95,27 @@ both. If both are specified, both conditions must be met for the requests to
...
+  protocol: ApplicationProtocol.HTTP,
+  targetGroupName: 'backend',
+  targets: [asg],
+  vpc: vpc,
+});
+
+const listenerRule = new ApplicationListenerRule(this, 'ListenerRule', {
+  listener,
+  priority: 10,
+  pathPatterns: ['/ok', '/path'],
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/README.md:112]: I think I'd like to see these fields `@deprecated` and a field added:

```
conditions?: Condition[];
```

(Obviously the @deprecated properties' code with go through the same code path).

Do this both for `ApplicationListenerRule` as well as for `listener.addTargets()` and `listener.addTargetGroups()`.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts [hunk 5]
```diff
@@ -169,7 +390,7 @@ export class ApplicationListenerRule extends cdk.Construct {
    */
   public readonly listenerRuleArn: string;
 
-  private readonly conditions: {[key: string]: string[] | undefined} = {};
+  private readonly conditions: {[key: string]: IListenerRuleCondition} = {};
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:393]: I think it's probably okay (and better), to treat the classes of conditions set using the `Condition` class and those set using key/value to be different. I'm not sure the overwrite behavior was ever a feature, more an implementation detail that was surfaced.

If you think the overwriting is a feature, then I guess feel free to retain it.

Otherwise, I'd have `conditions: Condition[]` and just append to it.

And then maybe there is also a `legacyConditions: Record<string, Condition>` for the conditions set using key and value, where we can have the overwrite behavior specifically for those situations where we had that as legacy behavior.

File: packages/@aws-cdk/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.expected.json (hunk 1 above)
Comment (karupanerura) [reply packages/@aws-cdk/aws-ecs-patterns/test/ec2/integ.multiple-application-load-balanced-ecs-service.expected.json:946]: I overlooked that problem. I'm agreed too.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts (hunk 2 above)
Comment (karupanerura) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:173]: That's better! I'll refactor as this!

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts (hunk 5 above)
Comment (karupanerura) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener-rule.ts:393]: > I'm not sure the overwrite behavior was ever a feature, more an implementation detail that was surfaced.

oh... I see.

I'll consider the design.

**Comment by karupanerura** [comment]
Thank you for the review! I'll fix these in a few days!

**Comment by karupanerura** [comment]
Oh... I'm sorry.
I accidentally rebased the branch in master synced local branch, which resulted in an unwanted commit log.
Do I need to recreate the Pull-Request?

**Comment by karupanerura** [comment]
@rix0rrr I'm recreate new PR #7848. could you continue to check this on the new PR?

**Comment by karupanerura** [comment]
@SomayaB @rix0rrr Please consider continuing the review on the re-created PR #7848. So close this PR.
